### PR TITLE
Fetch group info for users from DB when oktaMigrationEnabled is true

### PR DESCRIPTION
**Pull request by emyl3**
# BACKEND PULL REQUEST

## Related Issue

- Part of #7598

## Changes Proposed
- Regardless of `oktaMigrationEnabled` flag status, continue to update groups and roles in Okta
- When the `oktaMigrationEnabled` flag is set to `true`, read groups and roles from **DB**
- When the `oktaMigrationEnabled` flag is set to `false`, read groups and roles from **Okta**
 
Affected flows:
- adding org to queue (checking if the org is a duplicate)
- reprovisioning user
- updating user roles and privileges
- `users` graphql query (I don't think we are calling this on the frontend 🤔)
- verifying a pending organization
- getting org admin users
- confirmation modal when deleting a facility via site admin "Manage facility" tool (it displays how many users and patients are associated with only that facility)
Screenshot 2024-09-10 at 16 23 29

## Additional Information
- [Separate tickets](https://github.com/CDCgov/prime-simplereport/pull/8105/files#r1752181865) to address loading users with their status for the "Manage users" page

## Testing
dev2 - `oktaMigrationEnabled` - `true`
- [dev2 metabase link](https://dev2.simplereport.gov/metabase/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7ImRhdGFiYXNlIjoyLCJ0eXBlIjoicXVlcnkiLCJxdWVyeSI6eyJzb3VyY2UtdGFibGUiOjQxLCJqb2lucyI6W3sic3RyYXRlZ3kiOiJsZWZ0LWpvaW4iLCJhbGlhcyI6IkFwaSBVc2VyIE5vIFBoaSBWaWV3IC0gQXBpIFVzZXIiLCJjb25kaXRpb24iOlsiPSIsWyJmaWVsZCIsNDMxLHsiYmFzZS10eXBlIjoidHlwZS9VVUlEIn1dLFsiZmllbGQiLDQ0LHsiYmFzZS10eXBlIjoidHlwZS9VVUlEIiwiam9pbi1hbGlhcyI6IkFwaSBVc2VyIE5vIFBoaSBWaWV3IC0gQXBpIFVzZXIifV1dLCJzb3VyY2UtdGFibGUiOjExfSx7InN0cmF0ZWd5IjoibGVmdC1qb2luIiwiYWxpYXMiOiJBcGkgVXNlciBGYWNpbGl0eSAtIEFwaSBVc2VyIiwiY29uZGl0aW9uIjpbIj0iLFsiZmllbGQiLDQzMSx7ImJhc2UtdHlwZSI6InR5cGUvVVVJRCJ9XSxbImZpZWxkIiw0MjYseyJiYXNlLXR5cGUiOiJ0eXBlL1VVSUQiLCJqb2luLWFsaWFzIjoiQXBpIFVzZXIgRmFjaWxpdHkgLSBBcGkgVXNlciJ9XV0sInNvdXJjZS10YWJsZSI6NDJ9XSwiYnJlYWtvdXQiOltbImZpZWxkIiw0MzEseyJiYXNlLXR5cGUiOiJ0eXBlL1VVSUQifV0sWyJmaWVsZCIsNDMseyJiYXNlLXR5cGUiOiJ0eXBlL1RleHQiLCJqb2luLWFsaWFzIjoiQXBpIFVzZXIgTm8gUGhpIFZpZXcgLSBBcGkgVXNlciJ9XSxbImZpZWxkIiw0NDQseyJiYXNlLXR5cGUiOiJ0eXBlL1Bvc3RncmVzRW51bSJ9XSxbImZpZWxkIiw0MjgseyJiYXNlLXR5cGUiOiJ0eXBlL1VVSUQiLCJqb2luLWFsaWFzIjoiQXBpIFVzZXIgRmFjaWxpdHkgLSBBcGkgVXNlciJ9XSxbImZpZWxkIiw0NDUseyJiYXNlLXR5cGUiOiJ0eXBlL0RhdGVUaW1lIiwidGVtcG9yYWwtdW5pdCI6Im1pbnV0ZSJ9XSxbImZpZWxkIiw0NDMseyJiYXNlLXR5cGUiOiJ0eXBlL0RhdGVUaW1lIiwidGVtcG9yYWwtdW5pdCI6Im1pbnV0ZSJ9XSxbImZpZWxkIiw0MjkseyJiYXNlLXR5cGUiOiJ0eXBlL1VVSUQifV0sWyJmaWVsZCIsMzE3LHsiYmFzZS10eXBlIjoidHlwZS9UZXh0Iiwic291cmNlLWZpZWxkIjo0Mjl9XV0sIm9yZGVyLWJ5IjpbWyJkZXNjIixbImZpZWxkIiw0NDUseyJiYXNlLXR5cGUiOiJ0eXBlL0RhdGVUaW1lIiwidGVtcG9yYWwtdW5pdCI6Im1pbnV0ZSJ9XV1dfX0sImRpc3BsYXkiOiJ0YWJsZSIsInZpc3VhbGl6YXRpb25fc2V0dGluZ3MiOnt9LCJvcmlnaW5hbF9jYXJkX2lkIjpudWxsfQ==)

dev3 - `oktaMigrationEnabled` - `false`
- [dev3 metabase link](https://dev3.simplereport.gov/metabase/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7ImRhdGFiYXNlIjozLCJ0eXBlIjoicXVlcnkiLCJxdWVyeSI6eyJzb3VyY2UtdGFibGUiOjEyMSwiam9pbnMiOlt7InN0cmF0ZWd5IjoibGVmdC1qb2luIiwiYWxpYXMiOiJBcGkgVXNlciIsImNvbmRpdGlvbiI6WyI9IixbImZpZWxkIiwxMTY4LHsiYmFzZS10eXBlIjoidHlwZS9VVUlEIn1dLFsiZmllbGQiLDk0MCx7ImJhc2UtdHlwZSI6InR5cGUvVVVJRCIsImpvaW4tYWxpYXMiOiJBcGkgVXNlciJ9XV0sInNvdXJjZS10YWJsZSI6OTR9LHsiZmllbGRzIjoiYWxsIiwic3RyYXRlZ3kiOiJsZWZ0LWpvaW4iLCJhbGlhcyI6IkFwaSBVc2VyIEZhY2lsaXR5IC0gSW50ZXJuYWwiLCJjb25kaXRpb24iOlsiPSIsWyJmaWVsZCIsOTQwLHsiYmFzZS10eXBlIjoidHlwZS9VVUlEIiwiam9pbi1hbGlhcyI6IkFwaSBVc2VyIn1dLFsiZmllbGQiLDExNjMseyJiYXNlLXR5cGUiOiJ0eXBlL1VVSUQiLCJqb2luLWFsaWFzIjoiQXBpIFVzZXIgRmFjaWxpdHkgLSBJbnRlcm5hbCJ9XV0sInNvdXJjZS10YWJsZSI6MTIyfSx7ImZpZWxkcyI6ImFsbCIsInN0cmF0ZWd5IjoibGVmdC1qb2luIiwiYWxpYXMiOiJPcmdhbml6YXRpb24iLCJjb25kaXRpb24iOlsiPSIsWyJmaWVsZCIsMTE2Nix7ImJhc2UtdHlwZSI6InR5cGUvVVVJRCJ9XSxbImZpZWxkIiwxMDYxLHsiYmFzZS10eXBlIjoidHlwZS9VVUlEIiwiam9pbi1hbGlhcyI6Ik9yZ2FuaXphdGlvbiJ9XV0sInNvdXJjZS10YWJsZSI6Nzl9XSwiYnJlYWtvdXQiOltbImZpZWxkIiw5NDAseyJiYXNlLXR5cGUiOiJ0eXBlL1VVSUQiLCJqb2luLWFsaWFzIjoiQXBpIFVzZXIifV0sWyJmaWVsZCIsOTM5LHsiYmFzZS10eXBlIjoidHlwZS9UZXh0Iiwiam9pbi1hbGlhcyI6IkFwaSBVc2VyIn1dLFsiZmllbGQiLDEyMTMseyJiYXNlLXR5cGUiOiJ0eXBlL1Bvc3RncmVzRW51bSJ9XSxbImZpZWxkIiwxMTY1LHsiYmFzZS10eXBlIjoidHlwZS9VVUlEIiwiam9pbi1hbGlhcyI6IkFwaSBVc2VyIEZhY2lsaXR5IC0gSW50ZXJuYWwifV0sWyJmaWVsZCIsMTIxNCx7ImJhc2UtdHlwZSI6InR5cGUvRGF0ZVRpbWUiLCJ0ZW1wb3JhbC11bml0IjoibWludXRlIn1dLFsiZmllbGQiLDEyMTIseyJiYXNlLXR5cGUiOiJ0eXBlL0RhdGVUaW1lIiwidGVtcG9yYWwtdW5pdCI6Im1pbnV0ZSJ9XSxbImZpZWxkIiwxMDYwLHsiYmFzZS10eXBlIjoidHlwZS9UZXh0Iiwiam9pbi1hbGlhcyI6Ik9yZ2FuaXphdGlvbiJ9XSxbImZpZWxkIiwxMDYxLHsiYmFzZS10eXBlIjoidHlwZS9VVUlEIiwiam9pbi1hbGlhcyI6Ik9yZ2FuaXphdGlvbiJ9XV19fSwiZGlzcGxheSI6InRhYmxlIiwidmlzdWFsaXphdGlvbl9zZXR0aW5ncyI6e319) to see roles and facilities at a glance

### CODE REVIEW

File: backend/src/main/java/gov/cdc/usds/simplereport/service/ApiUserService.java [hunk 1]
```diff
@@ -597,10 +596,17 @@ public UserInfo getCurrentUserInfo() {
...
+    } else {
+      final Set<String> orgUserEmails = _oktaRepo.getAllUsersForOrganization(org);
+      usersInOrg = _apiUserRepo.findAllByLoginEmailInOrderByName(orgUserEmails);
+    }
+    return usersInOrg;
   }
 
+  // To be addressed in #8108
   @AuthorizationConfiguration.RequirePermissionManageUsers
   public List<ApiUserWithStatus> getUsersAndStatusInCurrentOrg() {
```
Comment (emyl3) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/ApiUserService.java:611]: ⚠️ holding off on updating this until this is addressed #8103

Then, this will be updated in #8108

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java [hunk 2]
```diff
@@ -364,7 +374,19 @@ public String verifyOrganizationNoPermissions(String externalId) {
...
-    return oktaRepository.activateOrganizationWithSingleUser(org);
+    if (featureFlagsConfig.isOktaMigrationEnabled()) {
+      Optional<ApiUser> orgAdmin =
+          dbAuthorizationService.getOrgAdminUsers(org).stream().findFirst();
+
+      if (orgAdmin.isPresent()) {
+        String orgAdminEmail = orgAdmin.get().getLoginEmail();
+        return oktaRepository.activateUser(orgAdminEmail);
+      } else {
+        throw new IllegalStateException("Organization does not have any org admins.");
```
Comment (emyl3) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java:385]: ⚠️ This `else` block changes the behavior of this method

Previously, even if the org had [no admin users this method would succeed](https://github.com/CDCgov/prime-simplereport/pull/8105/files#diff-92fa7f1f6942c235dcf316810541a74e42e432f1798b75f749fd4a273891c4f4R335)

However, with this implementation if an org does not have an admin it will [throw an exception](https://github.com/CDCgov/prime-simplereport/pull/8105/files#diff-92fa7f1f6942c235dcf316810541a74e42e432f1798b75f749fd4a273891c4f4R348).

Let me know if you think this should continue to have the same behavior as before. 🤔

File: backend/src/main/java/gov/cdc/usds/simplereport/service/DbAuthorizationService.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+  /**
+   * Fetches a count of ApiUsers that have permission to access the one defined facility and do not
+   * have the ALL_FACILITIES and/or ADMIN roles
+   *
+   * @param facility - Facility to get count for
+   * @return Integer - count of ApiUsers
+   */
+  public Integer getUserWithSingleFacilityAccessCount(Facility facility) {
+    List<ApiUser> users = _userRepo.findAllBySingleFacilityAccess(facility);
```
Comment (emyl3) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/DbAuthorizationService.java:52]: ⚠️ Let me know if this makes sense to generate this count 🙏

File: backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/AccountRequestController.java [hunk 4]
```diff
@@ -156,6 +163,19 @@ private String checkForDuplicateOrg(String organizationName, String state, Strin
     return String.join("-", organizationName, state);
   }
 
+  private List<String> getOrgAdminUserEmails(Organization org) {
+    List<String> adminUserEmails;
+    if (_ffc.isOktaMigrationEnabled()) {
+      adminUserEmails =
+          _das.getOrgAdminUsers(org).stream()
```
Comment (mpbrown) [inline backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/AccountRequestController.java:170]: nit: when looking over this code I had to scroll up further to see what `_das` was and I think the readability would feel more fluid renaming this to like `_dbAuthService` even though it's a longer variable name. Not that these all have to get changed right now but same feelings for `_ffc` to `_featureFlagConfig`, `_os` to `_orgService`, `_oqs` to `_orgQueueService`, `_aus` to `_apiUserService`, and `_es` to `_emailService`

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java (hunk 2 above)
Comment (mpbrown) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java:385]: I think this makes sense. Btw do we have any additional context on the comment on this method? Is this an old reference to some account request workflow that hasn't been implemented?

```
/**
   * This method is for verifying an organization after the Experian identity verification process.
   * It should not be used for any other purpose and once we move to the updated account request
   * workflow this should be removed.
   */
  @Transactional(readOnly = false)
  public String verifyOrganizationNoPermissions(String externalId) {
```

File: backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/AccountRequestController.java [hunk 5]
```diff
@@ -136,8 +142,9 @@ private String checkForDuplicateOrg(String organizationName, String state, Strin
       Optional<Organization> duplicateOrg =
           potentialDuplicates.stream().filter(o -> o.getExternalId().startsWith(state)).findFirst();
       if (duplicateOrg.isPresent()) {
-        if (_oktaRepo.fetchAdminUserEmail(duplicateOrg.get()).stream()
-            .anyMatch(Predicate.isEqual(email))) {
+        List<String> adminUserEmails = getOrgAdminUserEmails(duplicateOrg.get());
+
+        if (adminUserEmails.stream().anyMatch(Predicate.isEqual(email))) {
```
Comment (mpbrown) [inline backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/AccountRequestController.java:147]: Do we need to handle or otherwise flag if the org doesn't have any admins? (was thinking about no admin users after [this comment](https://github.com/CDCgov/prime-simplereport/pull/8105/files#r1752626902))

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java [hunk 6]
```diff
@@ -469,8 +491,15 @@ public FacilityStats getFacilityStats(@Argument UUID facilityId) {
         this.getFacilityById(facilityId)
             .orElseThrow(() -> new IllegalGraphqlArgumentException("Facility not found."));
 
+    Integer userWithSingleFacilityAccess;
```
Comment (mpbrown) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java:494]: nit: should this be users plural? Same with the method `dbAuthorizationService.getUserWithSingleFacilityAccessCount`?

**Review by mpbrown — COMMENTED** [review]
Looks really great overall! Awesome work on this Elisa!

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java (hunk 2 above)
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java:385]: That's a good question... Let me check that commit history in see if I can uncover anything... 🕵️

File: backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java [hunk 7]
```diff
@@ -30,4 +33,37 @@ public interface ApiUserRepository extends EternalSystemManagedEntityRepository<
 
   @Query(BASE_QUERY + " and loginEmail IN :emails" + NAME_ORDER)
   List<ApiUser> findAllByLoginEmailInOrderByName(Collection<String> emails);
+
+  String API_USER_ROLE_LEFT_JOIN = " LEFT JOIN api_user_role aur ON aur.apiUser = e";
+  String BY_ORG_AND_UNDELETED_USER = " WHERE aur.organization = :org AND e.isDeleted = false";
+  String JOINED_NAME_ORDER =
```
[the remark this answers is not in the document]
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java:39]: Great question! Since I am using a `LEFT JOIN` I need to specify that I want to order by the `ApiUser`'s `nameInfo` and `internalId` since the one you mentioned `NAME_ORDER` doesn't make that distinction.

Let me know if there's a nicer way to write this!! 👀 Thank you!!

File: backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/AccountRequestController.java (hunk 4 above)
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/AccountRequestController.java:170]: OOO good call out -- I can clean this up!

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java (hunk 2 above)
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java:385]: This is the PR that introduced this change: https://github.com/CDCgov/prime-simplereport/pull/2096
I did a search in the repo and couldn't find any mention of a new or updated account request flow 🤔 so I'm not entirely sure what that comment is referencing....

File: backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/AccountRequestController.java (hunk 5 above)
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/AccountRequestController.java:147]: I think it's OK in this context because we are trying to check if the to-be-created new org might be a duplicate of another org. If there are no org admins in that duplicate org, it would throw the exception on line 155. Let me know what you think!

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java (hunk 6 above)
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java:494]: OO nice catch! I'll address that!

File: backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/AccountRequestController.java (hunk 4 above)
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/api/accountrequest/AccountRequestController.java]: Done!

File: backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java (hunk 6 above)
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/OrganizationService.java]: Done!

**Review by bobbywells52 — DISMISSED** [review]
Tested in dev2 and dev3 and is working as excepted -- LGTM! Thank you for you continued heavy effort on this big lift  🙌🏻

**Comment by emyl3** [comment]
@mpbrown @mehansen @DanielSass Ready for re-review! Changes have been redeployed to dev2 and dev3 respectively

**Review by mpbrown — APPROVED** [review]
LGTM on dev2 and dev3! Great work on this Elisa!

File: backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java (hunk 7 above)
Comment (mehansen) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java:39]: nit: I think these would fit better and be easier to find declared at the top of the class with the other `NAME_ORDER` string

[the remark this answers is not in the document]
Comment (mehansen) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java:39]: I wonder if `NAME_ORDER` could be updated to specify the entity and we could just use that one everywhere instead of having two where one is just more explicit

File: backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java [hunk 8]
```diff
@@ -30,4 +33,37 @@ public interface ApiUserRepository extends EternalSystemManagedEntityRepository<
...
+  @Query(
+      value =
+          "from #{#entityName} e"
+              + API_USER_ROLE_LEFT_JOIN
+              + " LEFT JOIN api_user_facility auf ON auf.apiUser = e"
+              + " WHERE auf.facility = :facility"
+              + " AND e.isDeleted = false"
+              + " AND (aur.role = 'USER' OR aur.role = 'ENTRY_ONLY')"
+              + " GROUP BY e")
+  List<ApiUser> findAllBySingleFacilityAccess(Facility facility);
```
Comment (mehansen) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java:68]: I think the way you use this query to get all the users and then check for facility count is great but my nit is to change this method name since this technically doesn't do the "single facility" filtering yet

or make this method something like `findAllByFacilityAndRoles(Facility facility, Collection<OrganizationRole> roles)` and pass in the specific roles

File: backend/src/main/java/gov/cdc/usds/simplereport/service/DbAuthorizationService.java (hunk 3 above)
Comment (mehansen) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/DbAuthorizationService.java:52]: I think this will fix this bug https://github.com/CDCgov/prime-simplereport/issues/7873 🤩

**Review by mehansen — APPROVED** [review]
looks great! I have some non-blocking naming nits but they are more personal preference so still approving

File: backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java:39]: OOO good idea 😅 Will change that!

Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java:39]: Will do! Thank you!

File: backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java (hunk 8 above)
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java:68]: Good call! I'll change that!

File: backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java]: Done! thank you!

Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java]: Done!

File: backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java (hunk 8 above)
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/db/repository/ApiUserRepository.java]: Done!

File: backend/src/main/java/gov/cdc/usds/simplereport/service/ApiUserService.java [hunk 9]
```diff
@@ -829,4 +835,23 @@ private OrganizationRoles getOrgRolesFromDB(ApiUser apiUser) {
         _dbOrgRoleClaimsService.getOrganizationRoleClaims(apiUser);
     return _orgService.getOrganizationRoles(orgRoleClaims);
   }
+
+  private String getOrgExternalId(ApiUser apiUser) {
+    String orgExternalId;
+    if (_featureFlagsConfig.isOktaMigrationEnabled()) {
+      Optional<Organization> org = apiUser.getOrganizations().stream().findFirst();
```
Comment (DanielSass) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/ApiUserService.java:842]: Do we have users with more than one organization?  I don't think we do, but do we know that for certain / are there guardrails in place to ensure that can't happen?

**Review by DanielSass — APPROVED** [review]
Everything looks good, thanks for pairing with me on the review @emyl3. 
I think the only issues we noted were specific corner cases we need to make sure get included in the test plan.

Thanks for all your work on this

**Review by mehansen — APPROVED** [review]

**Review by mpbrown — APPROVED** [review]
